incubator/hnc/pkg/validators/config.go

```diff
@@ -0,0 +1,182 @@
+package validators
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
+)
+
+// ConfigServingPath is where the validator will run. Must be kept in sync with the
+// kubebuilder markers below.
+const (
+	ConfigServingPath = "/validate-hnc-x-k8s-io-v1alpha1-hncconfigurations"
+)
+
+// Note: the validating webhook FAILS OPEN. This means that if the webhook goes down, all further
+// changes to the objects are allowed.
+//
+// +kubebuilder:webhook:path=/validate-hnc-x-k8s-io-v1alpha1-hncconfigurations,mutating=false,failurePolicy=fail,groups="hnc.x-k8s.io",resources=hncconfigurations,verbs=create;update;delete,versions=v1alpha1,name=hncconfigurations.hnc.x-k8s.io
+
+type Config struct {
+	Log     logr.Logger
+	client  validationClient
+	decoder *admission.Decoder
+}
+
+// validationClient checks if a resource exists. The check should typically be performed against the apiserver,
+// but need to be stubbed out during unit testing.
+type validationClient interface {
+	// ResourceExists takes a GVK, and returns true if the corresponding resource of the GVK exists in the apiserver.
+	ResourceExists(ctx context.Context, gvk schema.GroupVersionKind) (bool, error)
+}
+
+type set map[schema.GroupVersionKind]bool
+
+func (c *Config) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if isHNCServiceAccount(&req.AdmissionRequest.UserInfo) {
+		return allow("HNC SA")
+	}
+
+	if req.Operation == v1beta1.Delete {
+		if req.Name == "config" {
+			return deny(metav1.StatusReasonForbidden, "Deleting the 'config' object is forbidden")
```

Comment:
"Forbidden" usually means "the current user is not authorized to take that action," which isn't the case here. See https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#StatusReason for the possible reasons. Maybe `StatusReasonBadRequest`?

Reply:
Thanks for the link. From the documentation, IIUC `not authorized` is covered by `StatusReasonUnauthorized `  not `StatusReasonForbidden` :

> StatusReasonUnauthorized means the server can be reached and understood the request, but requires the user to present appropriate authorization credentials  in order for the action to be completed. 

`StatusReasonBadRequest` might also not appropriate. That code means the request does not make any sense. However, in our case, we understand the request. Here is the documentation for `StatusReasonBadRequest`:

>   StatusReasonBadRequest means that the request itself was invalid, because the request doesn't make any sense, for example deleting a read-only object. API calls that return BadRequest can never succeed.

`StatusReasonInvalid` is also not appropriate here. From the documentation, it is for create and update operations and it means there is invalid data in the request, which does not apply here because we are doing a delete operation and the request itself is valid. 

> StatusReasonInvalid means the requested create or update operation cannot be completed due to invalid data provided as part of the request. The client may
need to alter the request. 

IMO `StatusReasonForbidden` is appropriate here. It means the server understand the request but deny the request:

> StatusReasonForbidden means the server can be reached and understood the request, but refuses to take any further action.  It is the result of the server being configured to deny access for some reason to the requested resource by the client.

Reply:
sgtm